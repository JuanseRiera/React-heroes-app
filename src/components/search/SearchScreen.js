import React, { useMemo } from "react";
import { HeroCard } from "../heroes/HeroCard";
import { useForm } from "../../hooks/useForm";
import { useLocation } from "react-router-dom";
import queryString from "query-string";
import { getHeroesByName } from "../../selectors/getHeroesByName";

export const SearchScreen = ({ history }) => {
  const { search } = useLocation();
  const { q = "" } = queryString.parse(search);
  const { formState, changeInput } = useForm({ searchHero: q });

  const heroes = useMemo(() => getHeroesByName(q), [q]);
  const buscarSubmit = (e) => {
    e.preventDefault();
    history.push(`?q=${formState.searchHero}`);
  };

  return (
    <>
      <h2>Search Screen</h2>
      <hr />
      <div className="row mt-4">
        <div className="col-12 col-md-5 mb-3">
          <h4>Search Form</h4>
          <hr />

          <form onSubmit={buscarSubmit}>
            <input
              type="text"
              placeholder="Buscar heroe..."
              autoComplete="off"
              className="form-control"
              name="searchHero"
              value={formState.searchHero}
              onChange={changeInput}
            />

            <button type="submit" className="btn btn-info mt-2">
              Buscar
            </button>
          </form>
        </div>

        <div className="col-12 col-md-7">
          <h4>Heroes encontrados:</h4>
          <hr />

          {q === "" && <div className="alert alert-info">Busque un heroe</div>}

          {q !== "" && heroes.length === 0 && (
            <div className="alert alert-danger">No se encontro el heroe</div>
          )}

          <div className="no-gutters mb-5 animate__animated animate__bounceInUp">
            {heroes.map((hero) => {
              return <HeroCard key={hero.id} hero={hero} />;
            })}
          </div>
        </div>
      </div>
    </>
  );
};
