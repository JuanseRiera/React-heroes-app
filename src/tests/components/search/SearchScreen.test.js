import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe("Pruabs del search screen", () => {
  test("Debe de cargaar correctamente con valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search" component={() => <SearchScreen />} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find(".alert-info").text().trim()).toBe("Busque un heroe");
  });

  test("Debe mostrar a batman y le input con el valor del queryString", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route path="/search" component={() => <SearchScreen />} />
      </MemoryRouter>
    );

    expect(wrapper.find("img").exists()).toBeTruthy();
    expect(wrapper.find("input").prop("value")).toBe("batman");
  });

  test("Debe de mostrar un error si no se encuentra el heroe", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman122121"]}>
        <Route path="/search" component={() => <SearchScreen />} />
      </MemoryRouter>
    );

    expect(wrapper.find(".alert-danger").exists()).toBeTruthy();
    expect(wrapper.find(".alert-danger").text().trim()).toBe(
      "No se encontro el heroe"
    );
  });

  test("Debe de llamar el push del history", () => {
    const history = {
      push: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route
          path="/search"
          component={() => <SearchScreen history={history} />}
        />
      </MemoryRouter>
    );

    wrapper.find("input").simulate("change", {
      target: {
        name: "searchHero",
        value: "batman",
      },
    });

    wrapper.find("button").simulate("submit", {
      preventEvent: jest.fn(),
    });

    expect(history.push).toHaveBeenCalledWith("?q=batman");
  });
});
