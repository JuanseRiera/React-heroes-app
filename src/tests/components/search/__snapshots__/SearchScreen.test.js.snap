// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruabs del search screen Debe de cargaar correctamente con valores por defecto 1`] = `ReactWrapper {}`;
